src/solvers/petsc_dm_wrapper.C

```diff
@@ -16,39 +16,355 @@
 // Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 
 #include "libmesh/libmesh_common.h"
+#include "libmesh/petsc_macro.h"
 
 #ifdef LIBMESH_HAVE_PETSC
+#if !PETSC_VERSION_LESS_THAN(3,7,3)
 
 #include "libmesh/ignore_warnings.h"
 #include <petscsf.h>
+#include <petsc/private/dmimpl.h>
 #include "libmesh/restore_warnings.h"
 
 #include "libmesh/petsc_dm_wrapper.h"
 #include "libmesh/system.h"
+#include "libmesh/mesh.h"
 #include "libmesh/mesh_base.h"
+#include "libmesh/mesh_refinement.h"
+#include "libmesh/partitioner.h"
 #include "libmesh/dof_map.h"
 #include "libmesh/elem.h"
-#include "libmesh/dof_map.h"
 
 namespace libMesh
 {
 
+  //--------------------------------------------------------------------
+  // Functions with C linkage to pass to PETSc.  PETSc will call these
+  // methods as needed.
+  //
+  // Since they must have C linkage they have no knowledge of a namespace.
+  // We give them an obscure name to avoid namespace pollution.
+  //--------------------------------------------------------------------
+  extern "C"
+  {
+
+    //! Help PETSc create a subDM given a global dm when using fieldsplit
+#if PETSC_VERSION_LESS_THAN(3,9,0)
+    PetscErrorCode __libmesh_petsc_DMCreateSubDM(DM dm, PetscInt numFields, PetscInt fields[], IS *is, DM *subdm)
+#else
+    PetscErrorCode __libmesh_petsc_DMCreateSubDM(DM dm, PetscInt numFields, const PetscInt fields[], IS *is, DM *subdm)
+#endif
+    {
+      PetscErrorCode ierr;
+      // Basically, we copy the PETSc ShellCreateSubDM implementation,
+      // but also need to set the embedding dim and we also propagate
+      // the relevant function pointers to the subDM for GMG purposes.
+      // Since this is called by PETSc we gotta pull some of this info
+      // from the context in the DM.
+
+      // Get our context
+      void * ctx = NULL;
```

Comment:
I think you can just use `nullptr` here, it should interact seamlessly with C APIs AFAIK.

Reply:
It does, yeah.  Only difference between NULL and nullptr is typing, not representation.

Reply:
Cool that's exactly what I worried about, thanks for the info.